cmd/cloud_sql_proxy/internal/healthcheck/healthcheck.go

```diff
@@ -0,0 +1,133 @@
+// Copyright 2021 Google LLC All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package healthcheck tests and communicates the health of the Cloud SQL Auth proxy.
+package healthcheck
+
+import (
+	"context"
+	"errors"
+	"net"
+	"net/http"
+	"sync"
+
+	"github.com/GoogleCloudPlatform/cloudsql-proxy/logging"
+	"github.com/GoogleCloudPlatform/cloudsql-proxy/proxy/proxy"
+)
+
+const (
+	livenessPath = "/liveness"
+	readinessPath = "/readiness"
+)
+
+// Server is a type used to implement health checks for the proxy.
+type Server struct {
+	// started is a channel used to indicate whether the proxy has finished 
+	// starting up. The channel is open when startup is in progress and
+	// closed when startup is complete.
+	started chan struct{}
+	// once ensures that started can only closed once.
+	once sync.Once
```

Comment:
nit: I'd suggest making `once` a pointer - it can't be copied once it's been used